RecoHI/HiEvtPlaneAlgos/src/EvtPlaneProducer.cc

```diff
@@ -305,214 +459,160 @@ void EvtPlaneProducer::produce(edm::Event &iEvent, const edm::EventSetup &iSetup
         }
       }
     }
-
-    //
-    //Get flattening parameter file.
-    //
-    if (loadDB_) {
-      edm::ESHandle<RPFlatParams> flatparmsDB_;
-      iSetup.get<HeavyIonRPRcd>().get(flatparmsDB_);
-      LoadEPDB db(flatparmsDB_, flat);
-      if (!db.IsSuccess()) {
-        loadDB_ = kFALSE;
-      }
+  }
+  //
+  //Get flattening parameter file.
+  //
+  if (loadDB_ && hirpWatcher_.check(iSetup)) {
+    edm::ESHandle<RPFlatParams> flatparmsDB_;
+    iSetup.get<HeavyIonRPRcd>().get(flatparmsDB_);
+    LoadEPDB db(flatparmsDB_, flat);
+    if (!db.IsSuccess()) {
+      loadDB_ = kFALSE;
     }
-
-  }  //rp record change
-
+  }
   //
   //Get Centrality
   //
   int bin = 0;
+  int cbin = 0;
   if (loadDB_) {
-    edm::Handle<int> cbin_;
-    iEvent.getByToken(centralityBinToken, cbin_);
-    int cbin = *cbin_;
+    cbin = iEvent.get(centralityBinToken_);
     bin = cbin / CentBinCompression_;
   }
   //
   //Get Vertex
   //
-  int vs_sell = 0.;
-  float vzr_sell;
-  iEvent.getByToken(vertexToken, vertex_);
-  const reco::VertexCollection *vertices3 = nullptr;
-  if (vertex_.isValid()) {
-    vertices3 = vertex_.product();
-    vs_sell = vertices3->size();
-  }
-  if (vs_sell > 0) {
-    vzr_sell = vertices3->begin()->z();
-  } else
-    vzr_sell = -999.9;
-  //
+  //best vertex
+  const reco::Vertex &vtx = iEvent.get(vertexToken_)[0];
+  double bestvz = vtx.z();
+  double bestvx = vtx.x();
+  double bestvy = vtx.y();
+  double bestvzError = vtx.zError();
+  math::XYZPoint bestvtx(bestvx, bestvy, bestvz);
+  math::Error<3>::type vtx_cov = vtx.covariance();
+  //  Produce the EP regardless of vz position
+  //  if (bestvz < minvtx_ || bestvz > maxvtx_)
+  //    return;
+
   for (int i = 0; i < NumEPNames; i++)
     rp[i]->reset();
-  if (vzr_sell < minvtx_ or vzr_sell > maxvtx_)
-    return;
-
-  //calorimetry part
-
-  double tower_eta, tower_phi;
-  double tower_energyet, tower_energyet_e, tower_energyet_h;
-
-  iEvent.getByToken(caloToken, caloCollection_);
-
-  if (caloCollection_.isValid()) {
-    for (CaloTowerCollection::const_iterator j = caloCollection_->begin(); j != caloCollection_->end(); j++) {
-      tower_eta = j->eta();
-      tower_phi = j->phi();
-      tower_energyet_e = j->emEt();
-      tower_energyet_h = j->hadEt();
-      tower_energyet = tower_energyet_e + tower_energyet_h;
-      double minet = minet_;
-      double maxet = maxet_;
-      for (int i = 0; i < NumEPNames; i++) {
-        if (minet_ < 0)
-          minet = minTransverse[i];
-        if (maxet_ < 0)
-          maxet = maxTransverse[i];
-        if (tower_energyet < minet)
+  edm::Handle<edm::ValueMap<float>> chi2Map;
+  edm::Handle<edm::View<pat::PackedCandidate>> cands;
```

Comment:
```suggestion
  edm::Handle<pat::PackedCandidateCollection> cands;
```

Reply:
I tried to use PackedCandidateCollection, but the failed at compile time:
error: 'const class std::vector<pat::PackedCandidate>' has no member named 'ptrAt'
         track_.normalizedChi2 = (*chi2Map)[cands->ptrAt(i)];
Not sure how to fix it.

Reply:
`[cands->ptrAt(i)]`-> `[pat::PackedCandidateRef(cands, i)]` ?